### PR TITLE
Change TVL endpoint 

### PR DESCRIPTION
**Pull request by waynebruce0x**

### CODE REVIEW

File: defi/src/storeTvlInterval/computeTVL.ts [hunk 1]
```diff
@@ -1,98 +1,110 @@
...
-      body.timestamp = timestamp;
-    }
+  const burl = "https://coins.llama.fi/prices/";
+  const historical =
+    timestamp == "now" ? "current/" : `historical/${timestamp}/`;
+
+  for (let i = 0; i < Object.keys(balances).length; i += step) {
+    const coins = Object.keys(balances)
+      .slice(i, i + step)
+      .reduce((p, c) => p + `${c},`, "");
```
Comment (0xngmi) [inline defi/src/storeTvlInterval/computeTVL.ts:55]: .join(',') is much more efficient and readable than that reduce

appending string by string in a string that keeps growing is extremely slow, since it ends up with multiple memory allocations

something like this:
1. allocate memory for string1
2. do string = string1+string2, now you need more space so you allocate space for the new string, copy string1 there and then copy string2
3. do string += string3, same again: new memory allocated and whole old string is copied there

this is generally super slow since you need to be reallocating ever growing strings

idk what join() does but i guess it calculates the final length of the string, allocates it beforehand and then copies all the strings there

File: defi/src/storeTvlInterval/computeTVL.ts [hunk 2]
```diff
@@ -1,98 +1,110 @@
...
+}
+
+const confidenceThreshold = 0.5;
+const step = 40;
+export default async function (
+  balances: { [address: string]: string },
+  timestamp: "now" | number,
+  staleCoins: any
+) {
+  staleCoins; // can get rid of stalecoins type?
```
Comment (0xngmi) [inline defi/src/storeTvlInterval/computeTVL.ts:74]: why is this here lol, doesnt this do nothing?

File: defi/src/storeTvlInterval/computeTVL.ts [hunk 3]
```diff
@@ -1,98 +1,110 @@
...
-        }
-        const usdAmount = amount * price;
-        tokenBalances[symbol] = (tokenBalances[symbol] ?? 0) + amount;
-        usdTokenBalances[symbol] = (usdTokenBalances[symbol] ?? 0) + usdAmount;
-        usdTvl += usdAmount;
-      });
-    } else {
-      console.error(`Data for ${response.PK} is stale`);
-    }
+    Object.keys(response).forEach((address) => {
```
Comment (0xngmi) [inline defi/src/storeTvlInterval/computeTVL.ts:84]: i assume the point of removing stalecoins tracking is that now we already remove stale coins on coins backend but the big issue is that we dont track them there + cant track tvl impact there

File: defi/src/storeTvlInterval/computeTVL.ts [hunk 4]
```diff
@@ -1,98 +1,110 @@
...
+    }
+
+    const normalisedKey = key.startsWith("0x")
+      ? `ethereum:${key.toLowerCase()}`
+      : !key.includes(":")
+      ? `coingecko:${key.toLowerCase()}`
+      : key;
+
+    // sols case sensitive so no normalising
+    if (key == normalisedKey || key.startsWith("solana:")) return;
```
Comment (g1nt0ki) [inline defi/src/storeTvlInterval/computeTVL.ts:24]: we dont have much support for tezos, but same applies for tezos, case sensitive

File: defi/src/storeTvlInterval/computeTVL.ts [hunk 5]
```diff
@@ -1,98 +1,110 @@
...
-        usdTvl += usdAmount;
-      });
-    } else {
-      console.error(`Data for ${response.PK} is stale`);
-    }
+    Object.keys(response).forEach((address) => {
+      const data = response[address];
+      const balance: BigNumber = new BigNumber(balances[address]);
+
+      if (data == undefined)
```
Comment (0xngmi) [inline defi/src/storeTvlInterval/computeTVL.ts:88]: isnt this impossible?

File: defi/src/storeTvlInterval/computeTVL.ts [hunk 6]
```diff
@@ -1,98 +1,110 @@
...
+    if (+balances[key] === 0) {
+      delete balances[key];
+      return;
+    }
+
+    const normalisedKey = key.startsWith("0x")
+      ? `ethereum:${key.toLowerCase()}`
+      : !key.includes(":")
+      ? `coingecko:${key.toLowerCase()}`
+      : key;
```
Comment (g1nt0ki) [inline defi/src/storeTvlInterval/computeTVL.ts:21]: should this be `key.toLowerCase()`?

**Review by 0xngmi — COMMENTED** [review]
not very confident in my review of `coins/src/adapters/utils/database.ts` since im not familiar with that part of the codebase

File: defi/src/storeTvlInterval/computeTVL.ts [hunk 7]
```diff
@@ -1,98 +1,110 @@
...
+
+    // sols case sensitive so no normalising
+    if (key == normalisedKey || key.startsWith("solana:")) return;
+
+    balances[normalisedKey] =
+      normalisedKey in balances
+        ? (Number(balances[normalisedKey]) + Number(balances[key])).toString()
```
Comment (g1nt0ki) [inline defi/src/storeTvlInterval/computeTVL.ts:28]: 99% sure this is safe, but if we do `sdk.utils.sumSingleBalance(balances, normalisedKey, balances[key])` it uses bigNumber.js and takes care of edge cases/possible underflow/overflow?

File: defi/src/storeTvlInterval/computeTVL.ts (hunk 3 above)
Comment (0xngmi) [reply defi/src/storeTvlInterval/computeTVL.ts:84]: in my latest PR i use stalecoins to track how much TVL is associated with each stale coin: https://github.com/DefiLlama/defillama-server/pull/1101/files#diff-2ca08735ce63a32600edb7f03c852820e7433320a15f096e4264cb96eee5a717L70

Comment (0xngmi) [reply defi/src/storeTvlInterval/computeTVL.ts:84]: can probably figure out some other way since .timestamp is also provided by the new api, so not a deal breaker for this PR

File: defi/src/storeTvlInterval/computeTVL.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (g1nt0ki) [reply defi/src/storeTvlInterval/computeTVL.ts:24]: can probably add a method to sdk.utils ? like `normalizeBalances(balances)`

File: defi/src/storeTvlInterval/getAndStoreTvl.ts [hunk 8]
```diff
@@ -109,7 +109,7 @@ function mergeBalances(key:string, storedKeys:string[], balancesObject:tvlsObjec
     balancesObject[key] = {}
     storedKeys.map(keyToMerge=>{
       Object.entries(balancesObject[keyToMerge]).forEach((balance) => {
-        util.sumSingleBalance(balancesObject[key], balance[0], balance[1]);
+        util.sumSingleBalance(balancesObject[key], balance[0], BigNumber(balance[1] || '0').toFixed(0));
```
[the remark this answers is not in the document]
Comment (waynebruce0x) [reply defi/src/storeTvlInterval/getAndStoreTvl.ts:112]: was getting some 'invalid bignumber' errors before this change

File: coins/src/adapters/utils/database.ts [hunk 9]
```diff
@@ -138,22 +126,16 @@ async function getTokenAndRedirectDataDB(
...
-      );
-      if (selectedEntries.length == 0) {
-        return { PK: d.redirect, SK: d.SK };
-      } else {
-        return { PK: selectedEntries[0].redirect, SK: selectedEntries[0].SK };
-      }
-    });
+    // current redirects
+    const redirects: DbQuery[] = latestDbEntries
+      .map((d: DbEntry) => {
```
[the remark this answers is not in the document]
Comment (waynebruce0x) [reply coins/src/adapters/utils/database.ts:131]: I thought it'd be a good idea to start storing redirects only on SK=0 (along with decimals and symbols), meaning the redirect data is on latestDbEntries and not timedDbEntries. So I thought it'd be easier to map through latest instead of timed

File: coins/src/adapters/utils/database.ts [hunk 10]
```diff
@@ -245,44 +282,3 @@ export function filterWritesWithLowConfidence(allWrites: Write[]) {
...
 }
-function aggregateTokenAndRedirectData(reads: Read[]) {
-  const coinData: CoinData[] = reads
-    .map((r: Read) => {
-      const addressIndex: number = r.dbEntry.PK.indexOf(":");
-      const chainIndex = r.dbEntry.PK.indexOf("#");
-
-      let price =
-        r.redirect.length != 0 ? r.redirect[0].price : r.dbEntry.price;
-      if (price == undefined) price = -1;
```
[the remark this answers is not in the document]
Comment (waynebruce0x) [reply coins/src/adapters/utils/database.ts:256]: There shouldnt be but I can add the check back incase

File: defi/src/storeTvlInterval/computeTVL.ts (hunk 2 above)
Comment (waynebruce0x) [reply defi/src/storeTvlInterval/computeTVL.ts:74]: yeah I just put it there so it'd run

File: defi/src/storeTvlInterval/computeTVL.ts [hunk 11]
```diff
@@ -1,98 +1,110 @@
...
+      ? `ethereum:${key.toLowerCase()}`
+      : !key.includes(":")
+      ? `coingecko:${key.toLowerCase()}`
+      : key;
+
+    // sols case sensitive so no normalising
+    if (key == normalisedKey || key.startsWith("solana:")) return;
+
+    balances[normalisedKey] =
+      normalisedKey in balances
```
[the remark this answers is not in the document]
Comment (waynebruce0x) [reply defi/src/storeTvlInterval/computeTVL.ts:27]: this is handled on line 24

File: defi/src/storeTvlInterval/computeTVL.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (waynebruce0x) [reply defi/src/storeTvlInterval/computeTVL.ts:24]: like this? https://github.com/DefiLlama/defillama-sdk/pull/79

File: defi/src/storeTvlInterval/computeTVL.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (0xngmi) [reply defi/src/storeTvlInterval/computeTVL.ts:27]: ah yeah my bad, im stupid

File: defi/src/storeTvlInterval/computeTVL.ts (hunk 7 above)
Comment (0xngmi) [reply defi/src/storeTvlInterval/computeTVL.ts:28]: one issue is that some balances might be a coingecko id like
{
 bitcoin: 1.2
}

and `sumSingleBalance` runs `toFixed(0)` so what could happen is that 1.2 btc becomes 1 btc (we cut the decimals)

File: defi/src/storeTvlInterval/computeTVL.ts (hunk 3 above)
Comment (waynebruce0x) [reply defi/src/storeTvlInterval/computeTVL.ts:97]: had a go at reimplementing this - I guess the new endpoint wont return symbols and decimals for stale coins so we'd need to requery the DB, or just use the PK as reference

File: defi/src/storeTvlInterval/computeTVL.ts [hunk 12]
```diff
@@ -1,98 +1,110 @@
...
+    const normalisedKey = key.startsWith("0x")
+      ? `ethereum:${key.toLowerCase()}`
+      : !key.includes(":")
+      ? `coingecko:${key.toLowerCase()}`
+      : key;
+
+    // sols case sensitive so no normalising
+    if (key == normalisedKey || key.startsWith("solana:")) return;
+
+    balances[normalisedKey] =
```
[the remark this answers is not in the document]
Comment (waynebruce0x) [reply defi/src/storeTvlInterval/computeTVL.ts]: consolidated these, my implementation will need a small SDK change though

File: coins/src/adapters/utils/database.ts [hunk 13]
```diff
@@ -163,15 +145,70 @@ async function getTokenAndRedirectDataDB(
...
+        const addressIndex: number = latestDbEntry.PK.indexOf(":");
+        const chainIndex = latestDbEntry.PK.indexOf("#");
+
+        let confidence =
+          timedDbEntry != undefined && "confidence" in timedDbEntry
+            ? timedDbEntry.confidence
+            : undefined;
+
+        if ("redirect" in latestDbEntry) {
+          const redirect = timedRedirects.filter(
```
[the remark this answers is not in the document]
Comment (waynebruce0x) [reply coins/src/adapters/utils/database.ts]: Not 100% sure I understand correctly but I had a go and implementing this

File: coins/src/adapters/utils/database.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (0xngmi) [reply coins/src/adapters/utils/database.ts:256]: no need, i just wasnt sure about it and wanted to flag it

File: defi/src/storeTvlInterval/computeTVL.ts (hunk 3 above)
Comment (0xngmi) [reply defi/src/storeTvlInterval/computeTVL.ts:95]: requerying the db is not a problem imo, where can i find your re-implementation?
